test/Path/Gen.hs

```diff
@@ -19,66 +19,67 @@ import           Test.QuickCheck
 
 
 instance Validity (Path Abs File) where
-  isValid p@(Path fp)
-    =  FilePath.isAbsolute fp
-    && not (FilePath.hasTrailingPathSeparator fp)
-    && FilePath.isValid fp
-    && not (hasParentDir fp)
-    && (parseAbsFile fp == Just p)
+  validate p@(Path fp) =  check (FilePath.isAbsolute fp)                     "Path is absolute."
+                       <> check (not (FilePath.hasTrailingPathSeparator fp)) "Path does not have trailing seperator."
+                       <> check (FilePath.isValid fp)                        "Path is valid."
+                       <> check (not (hasParentDir fp))                      "Path does not have parent dir."
+                       <> check ((parseAbsFile fp == Just p))                "Path parses."
 
 instance Validity (Path Rel File) where
-  isValid p@(Path fp)
-    =  FilePath.isRelative fp
-    && not (FilePath.hasTrailingPathSeparator fp)
-    && FilePath.isValid fp
-    && fp /= "."
-    && not (hasParentDir fp)
-    && (parseRelFile fp == Just p)
+  validate p@(Path fp) =  check (FilePath.isRelative fp)                     "Path is relative."
+                       <> check (not (FilePath.hasTrailingPathSeparator fp)) "Path does not have trailing seperator."
+                       <> check (FilePath.isValid fp)                        "Path is valid."
+                       <> check (fp /= ".")                                  "Path is not '.' ."
+                       <> check (not (hasParentDir fp))                      "Path does not have parent dir."
+                       <> check ((parseRelFile fp == Just p))                "Path parses."
 
 instance Validity (Path Abs Dir) where
-  isValid p@(Path fp)
-    =  FilePath.isAbsolute fp
-    && FilePath.hasTrailingPathSeparator fp
-    && FilePath.isValid fp
-    && not (hasParentDir fp)
-    && (parseAbsDir fp == Just p)
-
+  validate p@(Path fp) =  check (FilePath.isAbsolute fp)               "Path is absolute."
+                       <> check (FilePath.hasTrailingPathSeparator fp) "Path has trailing seperator."
+                       <> check (FilePath.isValid fp)                  "Path is valid."
+                       <> check (not (hasParentDir fp))                 "Path does not have parent dir."
+                       <> check ((parseAbsDir fp == Just p))           "Path parses."
+                         
 instance Validity (Path Rel Dir) where
-  isValid p =
-    FilePath.isRelative fp
-    && FilePath.hasTrailingPathSeparator fp
-    && FilePath.isValid fp
-    && not (hasParentDir fp)
-    && (parseRelDir fp == Just p)
+  validate p =  check (FilePath.isRelative fp)               "Path is relative."
+             <> check (FilePath.hasTrailingPathSeparator fp) "Path has trailing seperator"
+             <> check (FilePath.isValid fp)                  "Path is valid."
```

Comment:
Valid according to `FilePath.isValid`, which is different because you're defining what it means for the path to be valid and it's not just this validity.